src/lib/common/include/sol-mainloop.h

```diff
@@ -64,11 +64,269 @@ extern "C" {
+ *    int status;
+ *
+ *    //Write the blob
+ *    blob = sol_ptr_vector_get_no_check(&handle->pending_blobs, 0);
+ *    status = my_device_write(handle->dev, blob->mem, blob->size);
+ *
+ *    // Data was written, inform the user
+ *    if (handle->tx_cb)
+ *       handle->tx_cb((void *)handle->user_data, handle, blob, status);
+ *    handle->pending_bytes -= blob->size;
+ *    sol_blob_unref(blob); //NOTE: that we unref the blob, not the user!
+ *
+ *    //Do we still have more data ?
+ *    sol_ptr_vector_del(&handle->pending_blobs, 0);
+ *    if (sol_ptr_vector_get_len(&handle->pending_blobs))
+ *       my_device_schedule_write(handle->dev, _write, handle);
+ * }
+ *
+ * SOL_API int
+ * my_stream_api_write(struct my_stream_api_handle *handle, struct sol_blob *blob)
+ * {
+ *    size_t total = handle->pending_bytes + blob->size;
+ *
+ *    //Tx_size was set and the total must not exceed tx_size
+ *    if (tx_size && total >= tx_size)
+ *       return -ENOSPC;
+ *
+ *    //Store the blob and ref it, since it will written later.
+ *    r = sol_ptr_vector_append(&handle->pending_blobs, blob);
+ *    SOL_INT_CHECK(r, < 0, r);
+ *    sol_blob_ref(blob);
+ *    handle->pending_bytes = total;
+ *
+ *    //Schedule the write operation
+ *    my_device_schedule_write(handle->dev, _write, handle);
+ *    return 0;
+ * }
+ * @endcode
+ *
+ * @subsection read Reading
+ *
+ * Since this is a stream, there's no direct way for a third party developer to request a read operation.
+ * In order to be able to read from the stream, the third party developer must provide the @c rx_cb durtion the
+ * stream configuration/creation. The @c rx_cb is provided during the stream creation, one must inform the user every time
+ * data is avaible to read.
+ * The @c rx_cb has the following signature:
+ *
+ * @code{.c}
+ * ssize_t (*rx_cb)(void *user_data, my_stream_api_handle *handle, const struct sol_buffer *buf);
+ * @endcode
+ *
+ * Note that the @c rx_cb returns a @c ssize_t, the returned value must be @c -errno if an error happened or the number
+ * of bytes consumed from @c buf (0 is valid). The consumed bytes will be removed from the @c buf.
+ *
+ * Below there's an example of the read implementation:
+ *
+ * @code{.c}
+ * //my_stream_api.c
+ *
+ * struct my_stream_api_handle {
+ *    Device *dev;
+ *    const void *user_data;
+ *    void (*tx_cb)(void *user_data, my_stream_api_handle *handle, struct sol_blob *blob, int status);
+ *    ssize_t (*rx_cb)(void *user_data, my_stream_api_handle *handle, const struct sol_buffer *buf);
+ *    struct sol_ptr_vector pending_blobs;
+ *    size_t tx_size;
+ *    size_t pending_bytes;
+ *    struct sol_timeout *read_timeout;
+ * };
+ *
+ * //Code...
+ *
+ * static bool
+ * _inform_user(void *data)
+ * {
+ *    struct my_stream_api_handle *handle = data;
+ *    ssize_t r;
+ *
+ *    //Inform the user
+ *    r = handle->rx_cb((void *)handle->user_data, handle, handle->rx);
+ *    //Remove the data.
+ *    if (r < 0)
```

Comment:
same as above, this can lead to crashes if the user closes/deletes the handle from the read callback.

Unlike above, the solution is not that simple since you need the return from user callback to change the handle if not deleted. Then add a `bool delete_me, in_use`. Before calling the user you `in_use = true`, then you check if `delete_me == true` and do the actual delayed deletion. On `delete` you check `in_use == true` and just mark `delete_me` and return.